File: leetcode-go/Q032-Longest-Valid-Parentheses/Q032-Longest-Valid-Parentheses.go
package Q032_Longest_Valid_Parentheses

func longestValidParentheses(s string) int {
	stack := []int{}
	for i := 0; i < len(s); i++ {
		if len(stack) != 0 && s[stack[len(stack)-1]] == '(' && s[i] == ')' {
			stack = stack[0 : len(stack)-1]
		} else {
			stack = append(stack, i)
		}
	}

	if len(stack) == 0 {
		return len(s)
	}
	if stack[0] != 0 {
		stack = append([]int{-1}, stack...)
	}
	if stack[len(stack)-1] != len(s)-1 {
		stack = append(stack, len(s))
	}

	ans := 0
	for i := 0; i < len(stack)-1; i++ {
		if stack[i+1]-stack[i]-1 > ans {
			ans = stack[i+1] - stack[i] - 1
		}
	}
	return ans
}

File: src/vmware/Q032_Longest_Valid_Parentheses.java
package vmware;

import java.util.Stack;

public class Q032_Longest_Valid_Parentheses {

    public int longestValidParentheses(String s) {
        if (s == null) return 0;
        Stack<Integer> stack = new Stack<>();
        int result = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                if (!stack.isEmpty() && s.charAt(stack.peek()) == '(') {
                    stack.pop();
                    int pos = stack.isEmpty() ? -1 : stack.peek();
                    result = Math.max(result, i - pos);
                } else {
                    stack.push(i);
                }
            }
        }
        return result;
    }

}
